libs/wingc/src/jsify.rs

```diff
@@ -1520,15 +1580,175 @@ impl<'a> FieldReferenceVisitor<'a> {
 
 				let prop = vec![Component {
 					expr: node,
+					symbol: property.clone(),
 					variable: var,
 					text: property.name.to_string(),
 				}];
 
-				let obj = Self::analyze_expr(&object);
+				let obj = self.analyze_expr(&object);
 				return [obj, prop].concat();
 			}
 
 			_ => vec![],
 		}
 	}
 }
+
+struct FreeVariableScanner {
+	bound_vars: Vec<Symbol>,
+	free_vars: Vec<Symbol>,
+}
+
+impl FreeVariableScanner {
+	pub fn new(globals: Vec<Symbol>) -> Self {
+		Self {
+			bound_vars: globals,
+			free_vars: Vec::new(),
+		}
+	}
+}
+
+impl Visit<'_> for FreeVariableScanner {
+	fn visit_reference(&mut self, node: &Reference) {
+		if let Reference::Identifier(ref x) = node {
+			// check if there is already a bound variable with the same name
+			if !self.bound_vars.contains(x) {
+				self.free_vars.push(x.clone());
+			}
+		};
+
+		return visit::visit_reference(self, node);
+	}
+
+	// invariant: adds net zero bound variables
+	fn visit_scope(&mut self, scope: &Scope) {
+		let old_bound_vars = self.bound_vars.clone();
+		visit::visit_scope(self, scope);
+		// remove the bound variables that were introduced in this scope
+		self.bound_vars = old_bound_vars;
+	}
+
+	fn visit_stmt(&mut self, stmt: &Stmt) {
+		match &stmt.kind {
+			// this statement introduces bound variables!
+			StmtKind::Bring {
+				module_name,
+				identifier,
+			} => {
+				self.visit_symbol(&module_name);
+
+				// bring cloud;
```

Comment:
Nit: Add `"` to indicate this is what we are matching (or some other way)

```suggestion
				// "bring cloud;"
```